compiler/rustc_hir_typeck/src/cast.rs

```diff
@@ -803,40 +805,129 @@ impl<'a, 'tcx> CastCheck<'tcx> {
+                            ty::Dyn,
+                        ));
+                        let dst_obj = tcx.mk_ty_from_kind(ty::Dynamic(
+                            tcx.mk_poly_existential_predicates(
+                                &dst_tty.without_auto_traits().collect::<Vec<_>>(),
+                            ),
+                            tcx.lifetimes.re_erased,
+                            ty::Dyn,
+                        ));
 
-        // vtable kinds must match
-        if fcx.tcx.erase_regions(cast_kind) == fcx.tcx.erase_regions(expr_kind) {
-            Ok(CastKind::PtrPtrCast)
-        } else {
-            Err(CastError::DifferingKinds)
+                        // `dyn Src: Unsize<dyn Dst>`, this checks for matching generics
+                        let cause = fcx.misc(self.span);
+                        let obligation = Obligation::new(
+                            tcx,
+                            cause,
+                            fcx.param_env,
+                            ty::TraitRef::new(
+                                tcx,
+                                tcx.require_lang_item(LangItem::Unsize, Some(self.span)),
+                                [src_obj, dst_obj],
+                            ),
+                        );
+
+                        fcx.register_predicate(obligation);
+
+                        // Check that `SrcAuto` is a superset of `DstAuto`.
+                        // Emit an FCW otherwise.
+                        let src_auto = src_tty.auto_traits().collect::<FxHashSet<_>>();
+                        let added = dst_tty
+                            .auto_traits()
+                            .filter(|trait_did| !src_auto.contains(trait_did))
+                            .collect::<Vec<_>>();
+
+                        if !added.is_empty() {
+                            tcx.emit_node_span_lint(
+                                lint::builtin::PTR_CAST_ADD_AUTO_TO_OBJECT,
+                                self.expr.hir_id,
+                                self.span,
+                                errors::PtrCastAddAutoToObject {
+                                    traits_len: added.len(),
+                                    traits: added
+                                        .into_iter()
+                                        .map(|trait_did| {
+                                            format!("`{}`", tcx.def_path_str(trait_did))
+                                        })
+                                        .join(", "),
+                                },
+                            )
+                        }
+
+                        // FIXME: ideally we'd maybe add a flag here, so that borrowck knows that
+                        //        it needs to borrowck this ptr cast. this is made annoying by the
+                        //        fact that `thir` does not have `CastKind` and mir restores it
+                        //        from types.
```

Comment:
I don't know what it means 😅 

When would we like to borrowck these ptr casts? And does that simply mean not erasing lifetimes?

Reply:
I think this can be safely removed, because in borrowck we can match on the types just as well...

> When would we like to borrowck these ptr casts? And does that simply mean not erasing lifetimes?

When casting pointers to trait objects we need to make sure that `*const dyn T<'a>`->`*const dyn T<'b>` requires `'a: 'b`.